File: Dynamic Programming/edit_distance.py
class Solution(object):
    def minDistance(self, word1, word2):
        """
        :type word1: str
        :type word2: str
        :rtype: int
        这道题得用 DP 的方法去做，但是如果不知道递推关系是很难开头的，递推关系如下：
        设置一个 dp m*n 的数组做记录，dp[i][j] 表示第一个字符串的前 i 个字符和第二个字符串的前 j 个字符的最少操作数
        然后这个 dp[i][j] 是如何推出来的是关键，首先如果 i，j 中有一个为 0 那 dp[i][j] 肯定就等于 i 或者 j 了
        其次，如果不为 0，那就要看 word1[i-1] 和 word2[j-1] (索引从0开始的) 是否相等，如果相等，则操作数跟 dp[i-1][j-1] 一致
        如果不等，这操作数在 dp[i-1][j] 删除操作， dp[i][j-1] 增加操作， dp[i-1][j-1] 替换操作，这三个操作中的最小者 + 1
        具体这三个数为什么是对应这三个操作可以画图找他们的关系，知道了递推关系之后就可以写代码了 
        核心代码里面为两层循环，所以时间复杂度为 O(mn)， 用到的 dp 也是 m*n 规模的，所以空间复杂度也是这个昂
        """

        len1 = len(word1)
        len2 = len(word2)

        dp = [[0] * (len1 + 1) for _ in range(len2 + 1)]

        # 初始化， i,j 有一个为 0 的情况
        for i in range(1, len1+1): 
            dp[0][i] = i

        for j in range(1, len2+1):
            dp[j][0] = j


        for i in range(1, len2+1):
            for j in range(1, len1+1):
                # 字符串索引从 0 开始
                if word1[j-1] == word2[i-1]:
                    dp[i][j] = dp[i-1][j-1]
                else:
                    dp[i][j] = min(dp[i-1][j], dp[i][j-1], dp[i-1][j-1]) + 1

        return dp[-1][-1]



s = Solution()
print(s.minDistance("horse", "ros"))

File: Dynamic Programming/EditDistance.java
class EditDistance {
    public int minDistance(String word1, String word2) {
        return helper(word1, word2, 0, 0);
    }

    private int helper(String word1, String word2, int index, int mod) {
        if (word1 == null || index >= word1.length()) 
            return mod + (word1 == null ? word2.length() : word2.length()-word1.length());
        else if (index >= word2.length()) 
            return mod + word1.length() - word2.length();

        if (word1.charAt(index) == word2.charAt(index)) {
            return helper(word1, word2, index+1, mod);
        } else {
            String changeWord = word1.substring(0, index) + word2.charAt(index) + word1.substring(index+1);
            String deleteWord = word1.substring(0, index) + word1.substring(index+1);

            return Math.min(helper(changeWord, word2, index+1, mod+1), helper(deleteWord, word2, index, mod+1));
        }
    }

    /**
     * dp：dp[i][j] 表示长度为 i 的字符串 word1 转换成长度为 j 的字符串 word2 的最小操作数
     * 当 word1[i] == word2[j] 此时 dp[i][j] = dp[i-1][j-1]
     * 当 word1[i] != word2[j]
     *      把 word1[i] 替换成 word2[j] 则 dp[i][j] = dp[i-1][j-1] + 1
     *      在 word1 后面插入一个 word2[j] 则 dp[i][j] = dp[i][j-1] + 1
     *      把 word1[i] 删除 则 dp[i][j] = dp[i-1][j] + 1
     * 
     *      dp[i][j] = min(dp[i-1][j-1], dp[i][j-1], dp[i-1][j])
     */
    public int minDistance1(String word1, String word2) {
        int m = word1.length(), n = word2.length();

        int[][] dp = new int[m+1][n+1];
        
        // 当 word1 的长度为 0 时，则一直做插入操作
        for (int i = 1; i <= n; i++) {
            dp[0][i] = dp[0][i-1] + 1;
        }

        // 当 word2 的长度为 0 时，则一直做删除操作
        for (int i = 1; i <= m; i++) {
            dp[i][0] = dp[i-1][0] + 1;
        }

        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                if (word1.charAt(i-1) == word2.charAt(j-1)) {
                    dp[i][j] = dp[i-1][j-1];
                } else {
                    dp[i][j] = Math.min(Math.min(dp[i-1][j-1], dp[i][j-1]), dp[i-1][j]) + 1;
                }
            }
        }

        return dp[m][n];
    }

    public int minDistance2(String word1, String word2) {
        int m = word1.length(), n = word2.length();

        int[] dp = new int[n+1];
        int pre = 0;

        for (int i = 1; i <= n; i++) dp[i] = dp[i-1] + 1;

        for (int i = 1; i <= m; i++) {
            for (int j = 0; j <= n; j++) {
                if (j == 0) {
                    pre = dp[0];
                    dp[0] += 1;
                } else {
                    int temp = pre;
                    pre = dp[j];
                    if (word1.charAt(i-1) == word2.charAt(j-1)) {
                        dp[j] = temp;
                    } else {
                        dp[j] = Math.min(Math.min(temp, dp[j]), dp[j-1]) + 1;
                    }
                }
            }
        }

        return dp[n];
    }

    public static void main(String[] args) {
        EditDistance solution = new EditDistance();
        int result = solution.minDistance2("horse", "ros");
        System.out.println(result);
    }
}